kedro/io/catalog_config_resolver.py

```diff
@@ -0,0 +1,261 @@
+
+import copy
+import logging
+import re
+from typing import Any, Dict
+
+from parse import parse
+
+from kedro.io.core import DatasetError
+
+Patterns = Dict[str, Dict[str, Any]]
+
+CREDENTIALS_KEY = "credentials"
+
+
+def _fetch_credentials(credentials_name: str, credentials: dict[str, Any]) -> Any:
+    """Fetch the specified credentials from the provided credentials dictionary.
+
+    Args:
+        credentials_name: Credentials name.
+        credentials: A dictionary with all credentials.
+
+    Returns:
+        The set of requested credentials.
+
+    Raises:
+        KeyError: When a data set with the given name has not yet been
+            registered.
+
+    """
+    try:
+        return credentials[credentials_name]
+    except KeyError as exc:
+        raise KeyError(
+            f"Unable to find credentials '{credentials_name}': check your data "
+            "catalog and credentials configuration. See "
+            "https://kedro.readthedocs.io/en/stable/kedro.io.DataCatalog.html "
+            "for an example."
+        ) from exc
+
+
+def _resolve_credentials(
+    config: dict[str, Any], credentials: dict[str, Any]
+) -> dict[str, Any]:
+    """Return the dataset configuration where credentials are resolved using
+    credentials dictionary provided.
+
+    Args:
+        config: Original dataset config, which may contain unresolved credentials.
+        credentials: A dictionary with all credentials.
+
+    Returns:
+        The dataset config, where all the credentials are successfully resolved.
+    """
+    config = copy.deepcopy(config)
+
+    def _resolve_value(key: str, value: Any) -> Any:
+        if key == CREDENTIALS_KEY and isinstance(value, str):
+            return _fetch_credentials(value, credentials)
+        if isinstance(value, dict):
+            return {k: _resolve_value(k, v) for k, v in value.items()}
+        return value
+
+    return {k: _resolve_value(k, v) for k, v in config.items()}
+
+
+def _resolve_dataset_config(
+    ds_name: str,
+    pattern: str,
+    config: Any,
+) -> Any:
+    """Resolve dataset configuration based on the provided pattern."""
+    resolved_vars = parse(pattern, ds_name)
+    # Resolve the factory config for the dataset
+    if isinstance(config, dict):
+        for key, value in config.items():
+            config[key] = _resolve_dataset_config(ds_name, pattern, value)
+    elif isinstance(config, (list, tuple)):
+        config = [_resolve_dataset_config(ds_name, pattern, value) for value in config]
+    elif isinstance(config, str) and "}" in config:
+        try:
+            config = config.format_map(resolved_vars.named)
+        except KeyError as exc:
+            raise KeyError(
+                f"Unable to resolve '{config}' from the pattern '{pattern}'. Keys used in the configuration "
+                f"should be present in the dataset factory pattern."
+            ) from exc
+    return config
+
+
+class DataCatalogConfigResolver:
```

Comment:
I would've called it simply `CatalogConfigResolver`, since there's no other catalogs in our codebase except `DataCatalog`(s).